### PR TITLE
Update `datasets.py`

### LINKED ISSUE #67: Add `fetch_pose_data()` function
## The problem
We currently have a [`fetch_pose_data_path()`](https://github.com/neuroinformatics-unit/movement/blob/1d4aba5e879a8f692eaf6d4dc86c3247e635fa40/movement/datasets.py#L53C1-L54C1) function in the `datasets.py` module, for fetching the path to sample pose datasets stored on [GIN](https://gin.g-node.org/neuroinformatics/movement-test-data). This has been useful, especially in testing the I/O functionalities of `movement`. 

However if one wants to load one of the sample datasets into movement, a two-step procedure is needed, for example:
```python
file_path = datasets.fetch_pose_data_path(
    "SLEAP_some_file.analysis.h5"
)
ds = load_poses.from_sleap_file(file_path, fps=50)
```

## Solution
The above two steps could be combined into one by adding an extra `fetch_pose_data()` function, that fetches the path, applies the appropriate loading function from `load_poses()` (depending on whether it's a SLEAP or a DLC file) and returns the loaded dataset.

## Considerations
We may need to add some metadata for each sample dataset, most crucially the fps (obtain from researchers, with permission), so that we can load each sample dataset with the right time resolution.

### PR DESCRIPTION
**Pull request by b-peri**
## Description

This PR makes several important changes to the current workflow for dealing with movement's sample data:

1. First, to make more explicit that they handle sample datasets, the `datasets` module and it's component functions have been renamed:
     - `datasets.py` is now `sample_data.py`.
     - `list_pose_data()` is now `list_sample_data()`.
     - `fetch_pose_data_path()` is now `fetch_sample_data_path`.
3. Second, to avoid needing to update _both_ GIN and GitHub any time new sample data is added to the repository, hard-coded filenames and `sha256sum` keys have been removed from `sample_data.py`. The download registry is now built dynamically using metadata stored on the GIN repository itself (`poses_files_metadata.yaml`). To this end, two internal functions which fetch this metadata file from GIN and load it into movement (`_download_metadata_file` and `_fetch_metadata`, respectively) have been added.
4. Finally, this PR adds `sample_data.fetch_sample_data`, a new user-facing function which enables the user to _fetch_ a sample dataset from the GIN repository and _load_ it as an `xarray.Dataset` object (with the correct FPS) in one line.

## References
Closes #67 

## Checklist:
- [x] The code has been tested locally
- [x] Tests have been added to cover all new functionality
- [x] The documentation has been updated to reflect any changes
- [x] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Comment by b-peri** [comment]
### **TO-DO**:

- Write tests
- Add sample metadata yaml to GIN.
- Fetch sample metadata from GIN and pass to `load_poses` function.

**Comment by niksirbi** [comment]
## Relevant notes from today's meeting

- We decided to keep hosting the metadata.yaml file on GIN - this will be the canonical place for information about our sample data
- This can be fetched by pooch and based on the file's contained info, we can create the data registry object with the video file names and hashes
- The hash of the metadata file itself will not be checked
- The goal is that we don't have to update the code repo every time we update the data repo
- Need to test the behaviour of pooch when either the metadata, or the data itself updates on GIN

File: movement/sample_data.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    ----------
+    filename : str
+        Name of the file to fetch.
+
+    Returns
+    -------
+    path : pathlib.Path
+        Path to the downloaded file.
+    """
+    return Path(SAMPLE_DATA.fetch(filename, progressbar=True))
```
Comment (niksirbi) [inline movement/sample_data.py:81]: Sorry for the scope creep, but I think we should catch the case in which the file is not in the registry, and print the list of available file names in the error message (for convenience).

I would do sth like:
```python
try:
    return Path(SAMPLE_DATA.fetch(filename, progressbar=True))
except ValueError as error:
    raise log_error(
        error, 
        f"File '{file_path}' is not in the registry. Valid filenames are: {list_sample_data()}"
    )
```

The type annotations and the tests should then be changed accordingly.

File: movement/sample_data.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+
+def list_sample_data() -> list[str]:
+    """Find available sample pose data in the *movement* data repository.
+
+    Returns
+    -------
+    filenames : list of str
+        List of filenames for available pose data."""
+    return list(SAMPLE_DATA.registry.keys())
```
Comment (niksirbi) [inline movement/sample_data.py:60]: this line is currently the only one in this module not covered by tests. If you take care of that, I think codecov will stop complaining in CI.

File: docs/source/getting_started.md [hunk 3]
```diff
@@ -114,27 +114,36 @@ You can also try movement out on some sample data included in the package.
...
+file_path = sample_data.fetch_sample_data_path("DLC_two-mice.predictions.csv")
 ```
 The first time you call this function, it will download the corresponding file
 to your local machine and save it in the `~/.movement/data` directory. On
 subsequent calls, it will simply return the path to that local file.
 
 You can feed the path to the `from_dlc_file` or `from_sleap_file` functions
 and load the data, as shown above.
+
+Alternatively, you can skip the `fetch_sample_data_path()` step and load the
```
Comment (niksirbi) [inline docs/source/getting_started.md:140]: This is due to my oversight, but we should also include `from_lp_file` in the line above this one.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -259,25 +259,28 @@ by the [German Neuroinformatics Node](https://www.g-node.org/).
...
 
-1. The `POSE_DATA` download manager object, which contains a list of stored files and their known hashes.
-2. The `list_pose_data()` function, which returns a list of the available files in the data repository.
-3. The `fetch_pose_data_path()` function, which downloads a file (if not already cached locally) and returns the local path to it.
+1. The `METADATA` variable, which contains detailed metadata (including filename, sha256 key, and source software) for every dataset
+ in the GIN repository.
+2. The `SAMPLE_DATA` download manager object.
+3. The `list_sample_data()` function, which returns a list of the available files in the data repository.
+4. The `fetch_sample_data_path()` function, which downloads a file (if not already cached locally) and returns the local path to it.
+5. The `fetch_sample_data()` function, which downloads and returns a given dataset as an `xarray.Dataset` directly.
```
Comment (niksirbi) [inline CONTRIBUTING.md:280]:
```suggestion
5. The `fetch_sample_data()` function, which downloads a file and loads it into movement directly, returning an `xarray.Dataset` object.
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -287,9 +290,10 @@ To add a new file, you will need to:
...
+5. Add your new files to `/movement-test-data/poses/`.
+6. Determine the sha256 checksum hash of each new file by running `sha256sum <filename>` in a terminal. Alternatively, you can use `pooch` to do this for you: `python -c "import pooch; hash = pooch.file_hash('/path/to/file'); print(hash)"`. If you wish to generate a text file containing the hashes of all the files in a given folder, you can use `python -c "import pooch; pooch.make_registry('/path/to/folder', 'sha256_registry.txt')`.
+7. Add metadata for your new files to `poses_files_metadata.yaml`, including their sha256 hashes.
+8. Commit your changes using `gin commit -m <message> <filename>`.
+9. Upload the committed changes to the GIN repository by running `gin upload`. Latest changes to the repository can be pulled via `gin download`. `gin sync` will synchronise the latest changes bidirectionally.
 
-You can also perform steps 3-6 via the GIN web interface, if you prefer to avoid using the CLI.
+You can also perform steps 3, 4, 8, and 9 via the GIN web interface, if you prefer to avoid using the CLI.
```
Comment (niksirbi) [inline CONTRIBUTING.md:299]: I think we should remove this now. Determining which exact steps can be done via the web interfaces gets harder. Let's have only one canonical way of doing this.

File: movement/sample_data.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    "https://gin.g-node.org/neuroinformatics/movement-test-data/raw/master"
+)
+
+# Save data in ¬/.movement/data
+DATA_DIR = Path("~", ".movement", "data").expanduser()
+# Create the folder if it doesn't exist
+DATA_DIR.mkdir(parents=True, exist_ok=True)
+
+# Fetch newest sample metadata
+Path.unlink(DATA_DIR / "poses_files_metadata.yaml", missing_ok=True)
```
[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/sample_data.py:29]: Also, here you first delete the existing file, and then you download the new one.
I'm going to be paranoid and say what if the internet connection is down? Then your pooch request will fail and you will end up with no metadata files whatsoever. I think the prudent thing to do would be the following:
- rename the old metadata file to sth like "temp_poses_file_metadata.yaml", for safekeeping
- try to download the newest file, and if that fails rename the temp file back to its original filename, and raise a warning (this can be done with a try/except clause)

[the remark this answers is not in the document]
Comment (niksirbi) [reply movement/sample_data.py:29]: if the above logic get's too complex, you can encapsulate it in a function (sth like `fetch_metadata`), and add a line like `metadata = fetch_metadata(file_name="poses_files_metadata.yaml")`, which will also solve your DRY problem.

**Review by niksirbi — APPROVED** [review]
Great job @b-peri ! 

Thanks a lot for your contribution, I think we are very likely to merge this next week. In fact the feature works as is, which is great :tada:

I left many comments, but most of them are nitpicky and a matter of taste, so don't hesitate to push back on points.

Regarding the complaining codecov, I think you also need to cover `list_sample_data` with tests. You didn't introduce this function, but because we renamed the file, codecov thinks it's a new un-covered code line, so we might as well use the chance to test it.

File: movement/sample_data.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    path=DATA_DIR,
+    base_url=f"{DATA_URL}",
+    registry={"poses_files_metadata.yaml": None},
+)
+METADATA_PATH = Path(
+    METADATA_POOCH.fetch("poses_files_metadata.yaml", progressbar=True)
+)
+
+with open(METADATA_PATH, "r") as sample_info:
+    METADATA = yaml.safe_load(sample_info)
```
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/sample_data.py:40]: Agree completely! Will do!

File: movement/sample_data.py [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+SAMPLE_REGISTRY = {file["file_name"]: file["sha256sum"] for file in METADATA}
```
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/sample_data.py:42]: You're right - well spotted. I'll change this.

File: tests/conftest.py [hunk 9]
```diff
@@ -8,16 +8,16 @@
...
 
 def pytest_configure():
     """Perform initial configuration for pytest.
     Fetches pose data file paths as a dictionary for tests."""
     pytest.POSE_DATA = {
-        file_name: fetch_pose_data_path(file_name)
+        file_name: fetch_sample_data_path(file_name)
         for file_name in [
             "DLC_single-wasp.predictions.h5",
             "DLC_single-wasp.predictions.csv",
```
[the remark this answers is not in the document]
Comment (b-peri) [reply tests/conftest.py:23]: Agree completely!

File: tests/test_unit/test_sample_data.py [hunk 10]
```diff
@@ -0,0 +1,24 @@
+"""Test suite for the load_poses module."""
```
[the remark this answers is not in the document]
Comment (b-peri) [reply tests/test_unit/test_sample_data.py:1]: Oops, missed that! Thanks!

File: movement/sample_data.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+# Fetch newest sample metadata
+Path.unlink(DATA_DIR / "poses_files_metadata.yaml", missing_ok=True)
+METADATA_POOCH = pooch.create(
+    path=DATA_DIR,
+    base_url=f"{DATA_URL}",
+    registry={"poses_files_metadata.yaml": None},
+)
+METADATA_PATH = Path(
+    METADATA_POOCH.fetch("poses_files_metadata.yaml", progressbar=True)
+)
```
[the remark this answers is not in the document]
Comment (b-peri) [reply movement/sample_data.py]: Hi Niko!

1. > Since it's only one file, I think we don't need need to create a `pooch` registry. `pooch` also provides funcitonality for retrieving a single file from a URL, sth like:
    > 
    > ```python
    > metadata_path = pooch.retrieve(
    >     url=f"{DATA_URL}/poses_files_metadata.yaml",
    >     known_hash=None
    > )
    > ```
    Good point. For some reason I was under the impression that it wasn't possible to change the download path from pooch's own (OS-specific) `pooch.os_cache`, so I initially opted for `pooch.create()` + `pooch.fetch()` approach, thinking that manually moving the file from the cache to `.movement/data` after `pooch.fetch()`'ing it would be equally long, and probably harder to follow. However, after having another look at the docs for `pooch.fetch()` I see that I was mistaken! I'll fix this right away.
    
2. > Note that I'm using lowercase here. You picked up my pattern of over-using capitalisation in this file.
    > 
    > How about we unly use capitalisation for `DATA_URL`, `DATA_DIR` (as these are constants that are unlikely to change) and for `SAMPLE_DATA` (to distinguish the download manager from the module itself). I think everything else can be lowercase, what do you think?
    All good! I was indeed just trying to keep a consistent style, but wasn't completely sure what needed to be capitalized and what didn't! I'll change this.

File: movement/sample_data.py (hunk 1 above)
Comment (b-peri) [reply movement/sample_data.py]: Hi Niko, I've now implemented the try block as you suggested, but I'm not sure I understand how the type annotations and tests would now need to be changed! The test for `fetch_sample_data`, for instance, already included a case for nonexistent files. Let me know if there's anything I still need to change/add.

Comment (niksirbi) [reply movement/sample_data.py]: I thinks this is actually fine as is!

**Comment by niksirbi** [comment]
Thanks @b-peri for implementing my review suggestions!
I pushed some final edits to this branch. Specifically,
- I rewrote the `fetch_metadata` function, somewhat altering its functionality. Now this function will first attempt to download the newest metadata. If that fails, it will load an older local metadata file (if that exists) and raise a warning about that. If all the above fails, then it will raise an error. In the end, it loads and returns the metadata contents (not the file path)
- I added new tests for the above-described functionality. This was a bit tricky (since it also involved mocking a download failure). The new tests also check the format of the metadata file itself (to catch mistakes such as missing fields and typos in field names).
-  I refactored `test_sampe_data.py` and added some `pytest` fixtures to reduce code repetition.

Pleas do check the changes I've made and let me know if you find any error or if something is not clear!
